### PR TITLE
Split by timings for parallel run

### LINKED ISSUE #193: unreliable integration tests
Integration tests should not be coupled with each other (such that we need to preserve order of how they're executed) because it breaks test isolation and parallelism.

Integration tests are also very time consuming and we should utilize `--split-by=timings`. https://circleci.com/blog/a-guide-to-test-splitting/

Notifying @souradeep-das @Boyuan-Chen

### PR DESCRIPTION
**Pull request by trangtran-enya**
[:clipboard: Add associated issues, tickets, docs URL here.](https://github.com/bobanetwork/boba/issues/193)

## Overview

Currently, we already execute testcases in parallel. However, testcases are divided based on filename instead of execution time. 
[Issue 139](https://github.com/bobanetwork/boba/issues/193)

## Changes

- Add `mocha-junit-reporter` to collect timing data
- Turn off `hardhat-gas-reporter` as it conflicts with `mocha-junit-reporter`. Need discussion?
- Turn on split by timings
- split the time it takes for all tests to run in half

## Testing

- integration tests

### CODE REVIEW

File: ops/up_local.sh [hunk 1]
```diff
@@ -35,30 +35,30 @@ if [[ $BUILD == 1 ]]; then
...
-    docker-compose build -- replica
+  docker-compose build --parallel --no-cache -- builder l2geth l1_chain
+  docker-compose build --parallel --no-cache -- deployer dtl batch_submitter relayer integration_tests
+  docker-compose build --no-cache -- boba_message-relayer-fast
+  docker-compose build --no-cache -- gas_oracle
+  docker-compose build --no-cache -- boba_deployer
+  docker-compose build --no-cache -- fraud-detector
+  docker-compose build --no-cache -- monitor
+  docker-compose build --no-cache -- verifier
+  docker-compose build --no-cache -- replica
```
Comment (InoMurko) [inline ops/up_local.sh]: how come caches are disabled?

File: integration-tests/hardhat.config.ts [hunk 2]
```diff
@@ -5,7 +5,8 @@ import '@nomiclabs/hardhat-ethers'
 import '@nomiclabs/hardhat-waffle'
 import 'hardhat-gas-reporter'
 
-const enableGasReport = !!process.env.ENABLE_GAS_REPORT
+// const enableGasReport = !!process.env.ENABLE_GAS_REPORT
+const enableGasReport = false
```
Comment (InoMurko) [inline integration-tests/hardhat.config.ts]: why doesn't this work?

File: ops/up_local.sh (hunk 1 above)
Comment (trangtran-enya) [reply ops/up_local.sh]: Testcases in the beginning failed, so I tried to run with image built from latest code.

File: integration-tests/hardhat.config.ts (hunk 2 above)
Comment (trangtran-enya) [reply integration-tests/hardhat.config.ts]: Looks like the result from GasReport overwrite Timing result, which is used for future run.

File: ops/up_local.sh (hunk 1 above)
Comment (trangtran-enya) [reply ops/up_local.sh]: In the beginning, this worked. All testcases passed. Now, some testcases using cross-chain message failed.

File: integration-tests/hardhat.config.ts (hunk 2 above)
Comment (InoMurko) [reply integration-tests/hardhat.config.ts]: Have you tried things like:
https://github.com/mochajs/mocha/pull/1360
https://www.npmjs.com/package/mocha-multi-reporters
?

Comment (trangtran-enya) [reply integration-tests/hardhat.config.ts]: No, not yet. Let me try. Thanks Ino

The GasReporter is this one, configured through Hard-Hat. Maybe we will meet some problem. But I think I can figure out.

**Review by souradeep-das — APPROVED** [review]
lgtm

File: integration-tests/test/boba-fee-payment.spec.ts [hunk 3]
```diff
@@ -1044,9 +1044,9 @@ describe('Boba Fee Payment Integration Tests', async () => {
         bobaBalance
       )
       await transferBackTx.wait()
-    })
+    }).retries(3)
```
Comment (wsdt) [inline integration-tests/test/boba-fee-payment.spec.ts:1047]: why do we have a retry here? Just retry three times until we don't have enough Boba tokens anymore?

**Review by wsdt — APPROVED** [review]
looks great! 🔥

File: integration-tests/test/boba-api.spec.ts [hunk 4]
```diff
@@ -68,8 +68,8 @@ describe('Boba API Tests', async () => {
       process.env.L2_BOBA_ADDRESS = L2Boba.address
     })
 
-    describe('Mainnet', async () => {
-      it('{tag:boba} should swap BOBA for ETH', async () => {
```
Comment (boyuan-chen) [inline integration-tests/test/boba-api.spec.ts:72]: Do we want to remove all tags?

Comment (InoMurko) [reply integration-tests/test/boba-api.spec.ts:72]: eventually 🗡️

File: integration-tests/test/boba-fee-payment.spec.ts (hunk 3 above)
Comment (InoMurko) [reply integration-tests/test/boba-fee-payment.spec.ts:1047]: flaky test or flaky infra!

**Review by InoMurko — APPROVED** [review]

**Pull request merged**